### PR TITLE
Cookie sync improvements

### PR DESCRIPTION
**Pull request by matiaskorhonen**
Use a `WKHTTPCookieStoreObserver` to sync the cookies from the webview to `NSHTTPCookieStorage`.

Most developers shouldn't notice any difference in behaviour but based on my testing this seems to be more reliable than the old way where the cookies were synced in the `didFinishNavigation` callback. This also fixes a race condition bug that we hit with our production app.

### CODE REVIEW

**Comment by matiaskorhonen** [comment]
FYI, rebased against master now, @TheAlmightyBob & @Titozzz

**Comment by TheAlmightyBob** [comment]
Can you explain any more about what unreliability or race conditions you were experiencing that this addresses?

**Comment by matiaskorhonen** [comment]
Essentially rapid changes to the cookies inside the WebView would cause the NSHTTPCookieStorage to go out of sync resulting in authentication errors. Switching to this observer based approach fixed the issue for us.

**Comment by allan-simon** [comment]
@matiaskorhonen  do you think it relates to one of these issues ? https://github.com/react-native-webview/react-native-webview/issues/1780  and/or https://github.com/react-native-webview/react-native-webview/issues/2228 ? (our app also has issue with session stored in cookie so any improvement on that side is more than welcomed :) )

**Comment by matiaskorhonen** [comment]
> do you think it relates to one of these issues ? #1780 and/or #2228

Possibly, though I can't be sure as we mostly rely on the sync working *from* the webview to the shared cookie store (and not the other way around). 

I can say that the patch in this PR has improved the reliability of the cookie sync in our production app…

File: apple/RNCWebView.m [hunk 1]
```diff
@@ -37,7 +37,7 @@ -(id)inputAccessoryView
   if (_webView == nil) {
     return nil;
   }
-
```
Comment (allan-simon) [inline apple/RNCWebView.m:40]: shouldn't the PR  don't touch extra whitespace to ease the size of the diff and increase chance of it being merged ?

Comment (allan-simon) [reply apple/RNCWebView.m:40]: @matiaskorhonen  ping ? I think removing the extra whitespace change will increase a lot the chance of it being reviewed/merged

**Comment by allan-simon** [comment]
@TheAlmightyBob did you get time to review this, now that the diff is much much shorter ? :angel:

File: apple/RNCWebView.m [hunk 2]
```diff
@@ -1559,7 +1565,7 @@ - (void)resetupScripts:(WKWebViewConfiguration *)wkWebViewConfig {
       if(!_incognito && !_cacheEnabled) {
         wkWebViewConfig.websiteDataStore = [WKWebsiteDataStore nonPersistentDataStore];
       }
-      [self syncCookiesToWebView:nil];
```
Comment (TheAlmightyBob) [inline apple/RNCWebView.m:1562]: Did you mean to remove the `syncCookiesToWebView` call here? That's not equivalent to the `addObserver` line

**Review by TheAlmightyBob — CHANGES_REQUESTED** [review]
Sorry for the delay, I was hoping that somebody with better understanding of cookies would review, but you're right that it's at least not a big change :)

**Review by TheAlmightyBob — APPROVED** [review]

**Pull request merged**